src/gui/plugins/apply_force_torque/ApplyForceTorque.cc

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (C) 2023 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <mutex>
+#include <vector>
+#include <string>
+
+#include <gz/gui/Application.hh>
+#include <gz/gui/MainWindow.hh>
+#include <gz/sim/gui/GuiEvents.hh>
+#include <gz/sim/Link.hh>
+#include <gz/sim/Model.hh>
+#include <gz/sim/World.hh>
+#include <gz/sim/EntityComponentManager.hh>
+#include <gz/sim/components/World.hh>
+#include <gz/sim/components/Name.hh>
+#include <gz/msgs/Utility.hh>
+#include <gz/msgs/entity_wrench.pb.h>
+#include <gz/plugin/Register.hh>
+#include <gz/transport/Node.hh>
+
+#include "ApplyForceTorque.hh"
+
+namespace gz
+{
+namespace sim
+{
+  class ApplyForceTorquePrivate
+  {
+    /// \brief Publish wrench messages
+    public: void PublishWrench(bool _applyForce, bool _applyTorque);
+
+    /// \brief Transport node
+    public: transport::Node node;
+
+    /// \brief Publisher for EntityWrench messages
+    public: transport::Node::Publisher pub;
+
+    /// \brief To synchronize member access
+    public: std::mutex mutex;
```

Comment:
Since everything here is happening from a single Qt thread, I don't think you need this mutex.

Reply:
alas, I was wrong about this. The `EntitiesSelected` and `DeselectAllEntities` can come from the main Qt thread (via the `EntityTree` plugin) as well as the rendering thread (via the `SelectEntities` plugin), so a mutex is necessary. Sorry about that :(